app/models/deploy_queue.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+class DeployQueue
+  include Singleton
+
+  STAGGER_INTERVAL = 10.seconds
+
+  class << self
+    delegate :process, :dequeue, :queued?, to: :instance
+  end
+
+  def process(job_execution, deploy)
+    @lock.synchronize do
+      if @queue.empty? && can_deploy?
+        send_to_job_queue(job_execution, deploy)
+      else
+        enqueue(job_execution, deploy)
+      end
+    end
+  end
+
+  def dequeue
+    @lock.synchronize do
+      if @queue.present? && can_deploy?
```

Comment:
same check from above again ...

Reply:
Are they the same? One is `A ∧ B` and the other is `¬A ∧ B`?